### ISSUE TITLE
Out-of-memory error on TPUs - scaling Bayesian CNNs

### ISSUE THREAD

**Issue by mjbajwa** [pyro-ppl/numpyro#1292]
Hi, 

I'm trying to reproduce some results of [this paper](https://arxiv.org/abs/2104.14421) using NumPyro and running into some memory issues on TPUs. I have a TPU v2-8 instance where I have been developing. I'm using a very small CNN to test, but even that seems to run into memory issues. As a test, I have initialized the TPU instance with with standard Flax examples in the Flax repo (with an almost equivalent CNN to what I'm using in current work) and they seem to run without an issue. I also have a very simple 2-layered fully-connected network via NumPyro that runs on the TPU without a problem. 

I've put together a minimum reproducible example below. Would be grateful for your help!

**Error trace (from Colab TPU instance, but the error is almost identical on my TPU instance)**:

```
tpu
8
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow_datasets/core/dataset_builder.py:598: get_single_element (from tensorflow.python.data.experimental.ops.get_single_element) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.data.Dataset.get_single_element()`.
WARNING:tensorflow:From /usr/local/lib/python3.7/dist-packages/tensorflow_datasets/core/dataset_builder.py:598: get_single_element (from tensorflow.python.data.experimental.ops.get_single_element) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.data.Dataset.get_single_element()`.
Batch shape:  (1, 32, 32, 3)
Output shape:  (1, 10)
(8,)
(3, 3, 3, 8)
(16,)
(3, 3, 8, 16)
(32,)
(1024, 32)
(10,)
(32, 10)
Total parameters:  34522
---------------------------------------------------------------------------
UnfilteredStackTrace                      Traceback (most recent call last)
<ipython-input-3-26694e43617d> in <module>()
----> 1 run_conv_bnn(50000, 100, 100, False)

76 frames
UnfilteredStackTrace: RuntimeError: FAILED_PRECONDITION: Dependency failed: Dependency failed: Could not allocate 1639972864 bytes in memory 0x0x0_HBM0; 648003584 bytes allocatable, 2486272000 bytes available

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

RuntimeError                              Traceback (most recent call last)
/usr/local/lib/python3.7/dist-packages/jax/_src/numpy/lax_numpy.py in deferring_binary_op(self, other)
   6583     if not isinstance(other, _accepted_binop_types):
   6584       return NotImplemented
-> 6585     return binary_op(self, other)
   6586   return deferring_binary_op
   6587 

RuntimeError: FAILED_PRECONDITION: Dependency failed: Dependency failed: Could not allocate 1639972864 bytes in memory 0x0x0_HBM0; 648003584 bytes allocatable, 2486272000 bytes available
```

**Code**:

```
import os
import flax
import flax.linen as nn
import jax
import jax.numpy as jnp
import jax.random as random
import jax.tools.colab_tpu
import numpy as np
import numpyro
import numpyro.distributions as dist
import tensorflow.compat.v2 as tf
import jax.numpy as jnp
import numpy as np
import tensorflow_datasets as tfds
import tqdm
from sklearn.preprocessing import LabelBinarizer
from numpyro.contrib.module import random_flax_module
from numpyro.infer import MCMC, NUTS, Predictive, init_to_feasible, init_to_value

jax.tools.colab_tpu.setup_tpu()

def load_cifar10_dataset(train_index=50000, flatten=False):

    # Training dataset

    train_ds = tfds.as_numpy(tfds.load(
        'cifar10', split=tfds.Split.TRAIN, batch_size=-1))
    train_ds = {'image': train_ds['image'].astype(jnp.float32) / 255.,
                'label': train_ds['label'].astype(jnp.int32)}
    y_train_all = LabelBinarizer().fit_transform(train_ds['label'])

    # Test dataset

    test_ds = tfds.as_numpy(
        tfds.load('cifar10', split=tfds.Split.TEST, batch_size=-1))
    test_ds = {'image': test_ds['image'].astype(jnp.float32) / 255.,
               'label': test_ds['label'].astype(jnp.int32)}
    y_test = LabelBinarizer().fit_transform(test_ds['label'])

    # Filter to the first 1000 images for configuration

    temp_ds = {}
    temp_ds['image'] = train_ds['image'][0:train_index]
    temp_ds['label'] = train_ds['label'][0:train_index]
    y_train = y_train_all[0:train_index]

    if flatten:

        # Flatten for dense neural networks only

        train_x = np.zeros(
            shape=(temp_ds['image'].shape[0], np.prod(temp_ds['image'].shape[1:])))
        test_x = np.zeros(
            shape=(test_ds['image'].shape[0], np.prod(test_ds['image'].shape[1:])))

        for i, im in tqdm.tqdm(enumerate(temp_ds['image'])):
            train_x[i, :] = im.flatten()

        for i, im in tqdm.tqdm(enumerate(test_ds['image'])):
            test_x[i, :] = im.flatten()

    else:

        train_x = train_ds['image']
        test_x = test_ds['image']

    return train_x, test_x, y_train, y_test, temp_ds, test_ds

def run_conv_bnn(train_index=50000, num_warmup=100, num_samples=100, gpu=False):

    # Administrative stuff

    print(jax.default_backend())
    print(jax.device_count())
    
    # Disable tensorflow from using GPU

    tf.enable_v2_behavior()
    
    if gpu:

        physical_devices = tf.config.list_physical_devices('GPU')
    
        try:
            # Disable first GPU
            tf.config.set_visible_devices(physical_devices[1:], 'GPU')
            logical_devices = tf.config.list_logical_devices('GPU')
            # Logical device was not created for first GPU
            assert len(logical_devices) == len(physical_devices) - 1
        except:
            pass

        # Enable JAX/NumPyro to use GPU

        os.environ["XLA_PYTHON_CLIENT_ALLOCATOR"] = "platform" 
        os.environ["XLA_PYTHON_CLIENT_MEM_FRACTION"]=".8"
        numpyro.set_platform("gpu")
        numpyro.set_host_device_count(11)
    
    np.random.seed(0)

    # Declare constants for easy checks

    TRAIN_IDX = train_index
    NUM_WARMUP = num_warmup
    NUM_SAMPLES = num_samples

    # Create keys for numpyro

    rng_key, rng_key_predict = random.split(random.PRNGKey(0))

    # Load CIFAR-10 datasets

    train_x, test_x, y_train, y_test, temp_ds, test_ds = load_cifar10_dataset(train_index=TRAIN_IDX, flatten=False)
    
    # Define model

    class CNN(nn.Module):
            
        @nn.compact
        def __call__(self, x):
            
            x = nn.Conv(features=8, kernel_size=(3, 3))(x)
            x = nn.swish(x)
            x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
            x = nn.Conv(features=16, kernel_size=(3, 3))(x)
            x = nn.swish(x)
            x = nn.avg_pool(x, window_shape=(2, 2), strides=(2, 2))
            x = x.reshape((x.shape[0], -1))  # flatten
            x = nn.Dense(features=32)(x)
            x = nn.swish(x)
            x = nn.Dense(features=10)(x)
            x = nn.softmax(x)
                
            return x
        

    def model(x, y):
        
        module = CNN()
        
        net = random_flax_module(
            "CNN", 
            module, 
            # prior = dist.Normal(0, 100),
            prior = {
            "Conv_0.bias": dist.Normal(0, 100), 
            "Conv_0.kernel": dist.Normal(0, 100), 
            "Conv_1.bias": dist.Normal(0, 100), 
            "Conv_1.kernel": dist.Normal(0, 100), 
            "Dense_0.bias": dist.Normal(0, 100), 
            "Dense_0.kernel": dist.Normal(0, 100), 
            "Dense_1.bias": dist.Normal(0, 100), 
            "Dense_1.kernel": dist.Normal(0, 100),
            },
            input_shape=(1, 32, 32, 3)
        )
                
        numpyro.sample("y_pred", dist.Multinomial(total_count=1, probs=net(x)), obs=y)


    # Initialize parameters 

    model2 = CNN()
    batch = train_x[0:1, ]  # (N, H, W, C) format
    print("Batch shape: ", batch.shape)
    variables = model2.init(jax.random.PRNGKey(42), batch)
    output = model2.apply(variables, batch)      
    print("Output shape: ", output.shape)
    init = flax.core.unfreeze(variables)["params"]

    # Create more reasonable initial values by sampling from the prior

    prior_dist = dist.Normal(0, 10)
    init_new = init.copy()
    total_params = 0

    for i, high in enumerate(init_new.keys()):
        for low in init_new[high].keys():
            print(init_new[high][low].shape)
            init_new[high][low] = prior_dist.sample(jax.random.PRNGKey(i), init_new[high][low].shape)
            
            # increment count of total_params
            layer_params = np.prod(np.array([j for j in init_new[high][low].shape]))
            total_params += layer_params

    print("Total parameters: ", total_params)

    # Initialize MCMC

    kernel = NUTS(model, init_strategy=init_to_value(values=init_new))
    # kernel = NUTS(model, init_strategy=init_to_feasible(), target_accept_prob=0.80)
    mcmc = MCMC(  
        kernel,
        num_warmup=NUM_WARMUP,
        num_samples=NUM_SAMPLES,
        num_chains=1,
        progress_bar=True,
    )

    # Run MCMC

    mcmc.run(rng_key, train_x, y_train)

    # for i in range(NUM_SAMPLES):
    #    mcmc.run(random.PRNGKey(i), temp_ds['image'], y_train)
    #    batches = [mcmc.get_samples()]
    #    mcmc._warmup_state = mcmc._last_state

    # mcmc.print_summary()

    ### Prediction Utilities

    # TODO:

    # 1) Accuracy metrics on test and train
    # 2) Trace plots for parameters, or summary of R_hats across multiple chains
    # 3) Parameter posterior statistics (R_hat, n_eff)

    # TODO: convert the train_preds to probabilities over class, averaged by uncertainties?

    # Train accuracy calculation

    train_preds = Predictive(model, mcmc.get_samples())(jax.random.PRNGKey(2), train_x, y=None)["y_pred"]
    train_preds_ave = jnp.mean(train_preds, axis=0)
    train_preds_index = jnp.argmax(train_preds_ave, axis=1)
    accuracy = (temp_ds["label"] == train_preds_index).mean()*100
    print("Train accuracy: ", accuracy)

    # Test accuracy calculation

    test_preds = Predictive(model, mcmc.get_samples())(jax.random.PRNGKey(2), test_x, y=None)["y_pred"]
    test_preds_ave = jnp.mean(test_preds, axis=0)
    test_preds_index = jnp.argmax(test_preds_ave, axis=1)
    accuracy = (test_ds["label"] == test_preds_index).mean()*100
    print("Test accuracy: ", accuracy)

    # all_samples = mcmc.get_samples()
    # plt.plot(all_samples["CNN/Conv_0.kernel"][:, 3,3,3,16], "o")
    # plt.plot(all_samples["CNN/Dense_0.kernel"][:, 10], "o")

    # mcmc.print_summary()

run_conv_bnn(50000, 100, 100, False)
```

**Comment by mjbajwa** [comment]
Perhaps this more detailed error trace on the TPU instance might also help. I hypothesize a leakage somewhere in `flax_module`, but I have no idea where. 

```
jamal@t1v-n-2d3cc9c8-w-0:~/b-cnns$ python3 run_cnn.py
2022-01-17 15:40:23.392621: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
tpu
8
2022-01-17 15:40:31.233788: W tensorflow/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcuda.so.1'; dlerror: libcuda.so.1: cannot open shared object file: No such file or directory
2022-01-17 15:40:31.233830: W tensorflow/stream_executor/cuda/cuda_driver.cc:269] failed call to cuInit: UNKNOWN ERROR (303)
WARNING:tensorflow:From /home/jamal/.local/lib/python3.8/site-packages/tensorflow_datasets/core/dataset_builder.py:622: get_single_element (from tensorflow.python.data.experimental.ops.get_single_element) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.data.Dataset.get_single_element()`.
WARNING:tensorflow:From /home/jamal/.local/lib/python3.8/site-packages/tensorflow_datasets/core/dataset_builder.py:622: get_single_element (from tensorflow.python.data.experimental.ops.get_single_element) is deprecated and will be removed in a future version.
Instructions for updating:
Use `tf.data.Dataset.get_single_element()`.
Batch shape:  (1, 32, 32, 3)
Output shape:  (1, 10)
(4,)
(3, 3, 3, 4)
(8,)
(3, 3, 4, 8)
(16,)
(512, 16)
(10,)
(16, 10)
Total parameters:  8786
tcmalloc: large alloc 5529681920 bytes == 0xdac10000 @  0x7f44b8e4e680 0x7f44b8e6f824 0x7f44b153ceb5 0x7f44b15556b5 0x7f44b15565c1 0x7f44b1556f72 0x7f44b15c4554 0x7f44b15c5c4f 0x7f44b15c62e4 0x7f3fd649108a 0x7f44b0b14c07 0x7f44b0a61eef 0x7f44b07c2d53 0x7f44b07ab846 0x5f5e79 0x5f6a46 0x50b4a7 0x5703e6 0x5696da 0x5f6403 0x5f6eb8 0x6663ad 0x5f57ee 0x56cbfb 0x5696da 0x5f6403 0x5f6f19 0x6663ad 0x5f57ee 0x56cbfb 0x5f6226
tcmalloc: large alloc 1843404800 bytes == 0xc2570000 @  0x7f44b8e4e680 0x7f44b8e6f824 0x7f44b153ceb5 0x7f44b15556b5 0x7f44b15565c1 0x7f44b1556f72 0x7f44b15c4554 0x7f44b15c5c4f 0x7f44b15c62e4 0x7f3fd6afa08a 0x7f44b0b14c07 0x7f44b0a61eef 0x7f44b07c2d53 0x7f44b07ab846 0x5f5e79 0x5f6a46 0x50b4a7 0x5703e6 0x5696da 0x5f6403 0x5f6eb8 0x6663ad 0x5f57ee 0x56cbfb 0x5696da 0x5f6403 0x5f6f19 0x6663ad 0x5f57ee 0x56cbfb 0x5f6226
tcmalloc: large alloc 5529681920 bytes == 0xdac10000 @  0x7f44b8e4e680 0x7f44b8e6f824 0x7f44b153ceb5 0x7f44b15556b5 0x7f44b15565c1 0x7f44b1556f72 0x7f44b15c4554 0x7f44b15c5c4f 0x7f44b15c62e4 0x7f3fd649108a 0x7f44b0b14c07 0x7f44b0a61eef 0x7f44b07c2d53 0x7f44b07ab846 0x5f5e79 0x5f6a46 0x50b4a7 0x5703e6 0x5696da 0x5f6403 0x5f6eb8 0x6663ad 0x5f57ee 0x56cbfb 0x5696da 0x5f6403 0x5f6f19 0x6663ad 0x5f57ee 0x56cbfb 0x5f6226
tcmalloc: large alloc 1843404800 bytes == 0x10b950000 @  0x7f44b8e4e680 0x7f44b8e6f824 0x7f44b153ceb5 0x7f44b15556b5 0x7f44b15565c1 0x7f44b1556f72 0x7f44b15c4554 0x7f44b15c5c4f 0x7f44b15c62e4 0x7f3fd6afa08a 0x7f44b0b14c07 0x7f44b0a61eef 0x7f44b07c2d53 0x7f44b07ab846 0x5f5e79 0x5f6a46 0x50b4a7 0x5703e6 0x5696da 0x5f6403 0x5f6eb8 0x6663ad 0x5f57ee 0x56cbfb 0x5696da 0x5f6403 0x5f6f19 0x6663ad 0x5f57ee 0x56cbfb 0x5f6226
tcmalloc: large alloc 3686555648 bytes == 0x142838000 @  0x7f44b8e4e680 0x7f44b8e6f824 0x7f44b153ceb5 0x7f44b15556b5 0x7f44b15565c1 0x7f44b1556f72 0x7f44b15c4554 0x7f44b15c5c4f 0x7f44b15c62e4 0x7f3fd91a3255 0x7f44b0b14c07 0x7f44b0a61eef 0x7f44b07c2d53 0x7f44b07ab846 0x5f5e79 0x5f6a46 0x50b4a7 0x5703e6 0x5696da 0x5f6403 0x5f6eb8 0x6663ad 0x5f57ee 0x56cbfb 0x5696da 0x5f6403 0x5f6f19 0x6663ad 0x5f57ee 0x56cbfb 0x5f6226
2022-01-17 15:41:08.534348: E external/org_tensorflow/tensorflow/compiler/xla/pjrt/pjrt_stream_executor_client.cc:2085] Execution of replica 0 failed: RESOURCE_EXHAUSTED: Failed to allocate request for 782.00MiB (819986432B) on device ordinal 0
Traceback (most recent call last):
  File "run_cnn.py", line 211, in <module>
    run_conv_bnn(args.train_index, args.num_warmup, args.num_samples, args.gpu)
  File "run_cnn.py", line 159, in run_conv_bnn
    mcmc.run(rng_key, train_x, y_train)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/mcmc.py", line 572, in run
    states_flat, last_state = partial_map_fn(map_args)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/mcmc.py", line 360, in _single_chain_mcmc
    init_state = self.sampler.init(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/hmc.py", line 696, in init
    init_params = self._init_state(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/hmc.py", line 642, in _init_state
    init_params, potential_fn, postprocess_fn, model_trace = initialize_model(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 618, in initialize_model
    (init_params, pe, grad), is_valid = find_valid_initial_params(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 374, in find_valid_initial_params
    (init_params, pe, z_grad), is_valid = _find_valid_params(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 360, in _find_valid_params
    _, _, (init_params, pe, z_grad), is_valid = init_state = body_fn(init_state)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 345, in body_fn
    pe, z_grad = value_and_grad(potential_fn)(params)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 227, in potential_energy
    log_joint, model_trace = log_density_(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 53, in log_density
    model_trace = trace(model).get_trace(*model_args, **model_kwargs)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/handlers.py", line 165, in get_trace
    self(*args, **kwargs)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/primitives.py", line 87, in __call__
    return self.fn(*args, **kwargs)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/primitives.py", line 87, in __call__
    return self.fn(*args, **kwargs)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/primitives.py", line 87, in __call__
    return self.fn(*args, **kwargs)
  [Previous line repeated 2 more times]
  File "run_cnn.py", line 115, in model
    numpyro.sample("y_pred", dist.Multinomial(total_count=1, probs=net(x)), obs=y)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/contrib/module.py", line 117, in apply_without_state
    return nn_module.apply({"params": params}, *args, **kwargs)
  File "run_cnn.py", line 80, in __call__
    x = nn.swish(x)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/_src/nn/functions.py", line 101, in silu
    return x * sigmoid(x)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/_src/nn/functions.py", line 88, in sigmoid
    return expit(x)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/_src/scipy/special.py", line 104, in <lambda>
    expit.defjvps(lambda g, ans, x: g * ans * (lax._const(ans, 1) - ans))
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/_src/numpy/lax_numpy.py", line 6553, in deferring_binary_op
    return binary_op(self, other)
jax._src.source_info_util.JaxStackTraceBeforeTransformation: RuntimeError: RESOURCE_EXHAUSTED: Failed to allocate request for 782.00MiB (819986432B) on device ordinal 0

The preceding stack trace is the source of the JAX operation that, once transformed by JAX, triggered the following exception.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "run_cnn.py", line 211, in <module>
    run_conv_bnn(args.train_index, args.num_warmup, args.num_samples, args.gpu)
  File "run_cnn.py", line 159, in run_conv_bnn
    mcmc.run(rng_key, train_x, y_train)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/mcmc.py", line 572, in run
    states_flat, last_state = partial_map_fn(map_args)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/mcmc.py", line 360, in _single_chain_mcmc
    init_state = self.sampler.init(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/hmc.py", line 696, in init
    init_params = self._init_state(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/hmc.py", line 642, in _init_state
    init_params, potential_fn, postprocess_fn, model_trace = initialize_model(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 618, in initialize_model
    (init_params, pe, grad), is_valid = find_valid_initial_params(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 374, in find_valid_initial_params
    (init_params, pe, z_grad), is_valid = _find_valid_params(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 360, in _find_valid_params
    _, _, (init_params, pe, z_grad), is_valid = init_state = body_fn(init_state)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 345, in body_fn
    pe, z_grad = value_and_grad(potential_fn)(params)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/_src/traceback_util.py", line 162, in reraise_with_filtered_traceback
    return fun(*args, **kwargs)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/_src/api.py", line 884, in value_and_grad_f
    g = vjp_py(np.ones((), dtype=dtype))
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/_src/api.py", line 2052, in _vjp_pullback_wrapper
    ans = fun(*args)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/interpreters/ad.py", line 123, in unbound_vjp
    arg_cts = backward_pass(jaxpr, reduce_axes, consts, dummy_args, cts)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/interpreters/ad.py", line 222, in backward_pass
    cts_out = get_primitive_transpose(eqn.primitive)(
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/interpreters/ad.py", line 554, in call_transpose
    out_flat = primitive.bind(fun, *all_args, **new_params)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/core.py", line 1618, in bind
    return call_bind(self, fun, *args, **params)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/core.py", line 1609, in call_bind
    outs = primitive.process(top_trace, fun, tracers, params)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/core.py", line 1621, in process
    return trace.process_call(self, fun, tracers, params)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/core.py", line 615, in process_call
    return primitive.impl(f, *tracers, **params)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 625, in _xla_call_impl
    out = compiled_fun(*args)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 960, in _execute_compiled
    out_bufs = compiled.execute(input_bufs)
jax._src.traceback_util.UnfilteredStackTrace: RuntimeError: RESOURCE_EXHAUSTED: Failed to allocate request for 782.00MiB (819986432B) on device ordinal 0

The stack trace below excludes JAX-internal frames.
The preceding is the original exception that occurred, unmodified.

--------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "run_cnn.py", line 211, in <module>
    run_conv_bnn(args.train_index, args.num_warmup, args.num_samples, args.gpu)
  File "run_cnn.py", line 159, in run_conv_bnn
    mcmc.run(rng_key, train_x, y_train)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/mcmc.py", line 572, in run
    states_flat, last_state = partial_map_fn(map_args)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/mcmc.py", line 360, in _single_chain_mcmc
    init_state = self.sampler.init(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/hmc.py", line 696, in init
    init_params = self._init_state(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/hmc.py", line 642, in _init_state
    init_params, potential_fn, postprocess_fn, model_trace = initialize_model(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 618, in initialize_model
    (init_params, pe, grad), is_valid = find_valid_initial_params(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 374, in find_valid_initial_params
    (init_params, pe, z_grad), is_valid = _find_valid_params(
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 360, in _find_valid_params
    _, _, (init_params, pe, z_grad), is_valid = init_state = body_fn(init_state)
  File "/home/jamal/.local/lib/python3.8/site-packages/numpyro/infer/util.py", line 345, in body_fn
    pe, z_grad = value_and_grad(potential_fn)(params)
  File "/home/jamal/.local/lib/python3.8/site-packages/jax/interpreters/xla.py", line 960, in _execute_compiled
    out_bufs = compiled.execute(input_bufs)
RuntimeError: RESOURCE_EXHAUSTED: Failed to allocate request for 782.00MiB (819986432B) on device ordinal 0
```

**Comment by fehiepsi** [comment]
Here are some of my thoughts:
+ The number of samples is small and the number of parameters is not large, so they are unlikely to cause the issue.
+ The data is quite large and you are performing Conv on top of all data. So the first convolution computation might be the main source of high memory. You might test it outside of numpyro context.
+ You might use [jax profiling](https://jax.readthedocs.io/en/latest/device_memory_profiling.html) to see where the high memory is required. You might want to test for a smaller number of data points, `x_train[:10000]`, `y_train[:10000]`
+ For debugging, you might want to set `NUTS(..., max_tree_depth=1)` to improve speed.
+ Softmax + Multinomial might be expensive and subject to numerical issues. You might want to do
```
    class CNN(nn.Module):
        @nn.compact
        def __call__(self, x):
            ...
            x = nn.Dense(features=10)(x)                
            return x

...
    numpyro.sample("y_pred", dist.Categorical(logits=net(x)), obs=y1)
```
where `y1 = np.argmax(y, axis=1)`

**Comment by mjbajwa** [comment]
@fehiepsi your second point is indeed correct. Looks like reducing the "batch size" for a fixed model works. I imagine there's some sort of a time-space trade-off implicit in the convolution operations in XLA, but it's not obvious how I might control that through Flax. I'll do some digging. I'll shortly upgrade to a TPU instance with higher memory (v3-8), so hopefully it lets me do a full batch run. Thank you also for the Categorical tip.

I'l close this issue for now.

**Comment by kaijennissen** [comment]
@mjbajwa I'm trying to fit a Bayesian CNN on the MNIST dataset using numpyro, and ran into the same problem. Did you manage to get reasonable results?

**Comment by mjbajwa** [comment]
@kaijennissen yes, so I was able to resolve the memory issue by reducing batch size from 50,000 to 25,000, but I've only been able to hit 30% test set accuracy with a variation of the network above on CIFAR-10 (using standard NUTS settings). I'm also working on a v3-8 TPU instance and have ~128GB of memory.

I don't love the idea of needing to reduce the batch size; ofcourse sub-sampling is inconsistent with HMC principles. I'm still investigating if the memory blow-up due to the convolution operation can be controlled via JAX

I'll add running the code on MNIST to my to-do list, and will write back here if I have any other observations.

**Comment by mjbajwa** [comment]
@fehiepsi Based on your ideas and others, I was able to run small-scale Bayesian Convolutional Nets using mostly NumPyro. Got pretty reasonable results (71% test accuracy). All computation was done locally on my RTX 2080. I found that TPUs (v2) were very slow here; perhaps it was linked to convolution compilation. I'm not sure that 

Leaving my code here for future refrence, happy to contribute as an example: https://github.com/mjbajwa/b-cnns/blob/main/run_cnn.py
